### PR TITLE
add complete week3

### PR DESCRIPTION
**Pull request by MarcRaeder**
add week3

### CODE REVIEW

File: week3/plurality/plurality.c [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (candidates[i].votes == maximum_vote)
+        {
+            printf("%s\n", candidates[i].name);
+        }      
+    }
+    
+    return;
+}
```
Comment (matthiasdanne) [inline week3/plurality/plurality.c]:
```suggestion
}

```

File: week3/plurality/plurality.c [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    print_winner();
+}
+
+bool vote(string name)
+{
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (strcmp(candidates[i].name, name) == 0)
```
Comment (matthiasdanne) [inline week3/plurality/plurality.c]: Add speaking variable

File: week3/runoff/runoff.c [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (candidates[i].votes == min)
+        {
+            candidates[i].eliminated = true;
+        }
+    }
+    
+    return;
+}
```
Comment (matthiasdanne) [inline week3/runoff/runoff.c]:
```suggestion
}

```

File: week3/runoff/runoff.c [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+int voter_count;
+int candidate_count;
+
+bool vote(int voter, int rank, string name);
+void tabulate(void);
+bool print_winner(void);
+int find_min(void);
+bool is_tie(int min);
+void eliminate(int min);
```
Comment (matthiasdanne) [inline week3/runoff/runoff.c]: Should be ordered alphabetically.

File: week3/runoff/runoff.c [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    
+    return;
+}
+
+// Print the winner of the election, if there is one
+bool print_winner(void)
+{
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (candidates[i].votes > voter_count/2)
```
Comment (matthiasdanne) [inline week3/runoff/runoff.c]:
```suggestion
        if (candidates[i].votes > voter_count / 2)
```
Can be a speaking variable like `candidateHasAbsoluteMajorityOfVotes`

File: week3/runoff/runoff.c [hunk 6]
```diff
@@ -0,0 +1,213 @@
+#include <cs50.h>
```
Comment (matthiasdanne) [inline week3/runoff/runoff.c:1]: This works and looks good, overall. 👍

You should remove existing comments and add speaking variables where possible.

**Review by matthiasdanne — CHANGES_REQUESTED** [review]

File: week3/tideman/tideman.c [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        }
+
+        if (found_source)
+        {
+            printf("%s\n", candidates[i]);
+            return;
+        }
+    }
+    return;
+}
```
Comment (matthiasdanne) [inline week3/tideman/tideman.c]:
```suggestion
}

```

File: week3/tideman/tideman.c [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+int pair_count;
+int candidate_count;
+
+bool vote(int rank, string name, int ranks[]);
+void record_preferences(int ranks[]);
+void add_pairs(void);
+void sort_pairs(void);
+void lock_pairs(void);
+void print_winner(void);
```
Comment (matthiasdanne) [inline week3/tideman/tideman.c]: Sort alphabetically

File: week3/tideman/tideman.c [hunk 9]
```diff
@@ -0,0 +1,226 @@
+#include <cs50.h>
+#include <stdio.h>
+#include <string.h>
+
+#define MAX 9
+
+int preferences[MAX][MAX];
+
+bool locked[MAX][MAX];
```
Comment (matthiasdanne) [inline week3/tideman/tideman.c]: Move the variable initialization into one block.

File: week3/tideman/tideman.c [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    if (winner == loser)
+    {
+        return true;
+    }
+    return false;
+}
+
+int main(int argc, string argv[])
```
Comment (matthiasdanne) [inline week3/tideman/tideman.c:30]: I would expect the main function at the top of the file.

File: week3/tideman/tideman.c [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+            }
+        }
+
+        if (found_source)
+        {
+            printf("%s\n", candidates[i]);
+            return;
+        }
+    }
+    return;
```
Comment (matthiasdanne) [inline week3/tideman/tideman.c]:
```suggestion
    }
    
    return;
```

File: week3/tideman/tideman.c [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+void lock_pairs(void)
+{
+    for (int i = 0; i < pair_count; i++)
+    {
+        if (!hasCycle(pairs[i].winner, pairs[i].loser))
+        {
+            locked[pairs[i].winner][pairs[i].loser] = true;
+        }
+    }
+    return;
```
Comment (matthiasdanne) [inline week3/tideman/tideman.c]:
```suggestion
    }
    
    return;
```

File: week3/tideman/tideman.c [hunk 13]
```diff
@@ -0,0 +1,226 @@
+#include <cs50.h>
```
Comment (matthiasdanne) [inline week3/tideman/tideman.c:1]: Aside from minor code style, this looks very good. 

What happens if all votes are evenly distributed? (like below). I haven't found anything on a 'tie' case in the task description.

Bildschirmfoto 2022-08-15 um 13 05 39

**Review by matthiasdanne — CHANGES_REQUESTED** [review]

File: week3/plurality/plurality.c (hunk 1 above)
Comment (stefanluth) [inline week3/plurality/plurality.c]: [EOF](https://de.wikipedia.org/wiki/End_of_File) hinzufuegen

File: week3/tideman/tideman.c [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+        return 1;
+    }
+
+    candidate_count = argc - 1;
+    if (candidate_count > MAX)
+    {
+        printf("Maximum number of candidates is %i\n", MAX);
+        return 2;
+    }
+    for (int i = 0; i < candidate_count; i++)
```
Comment (matthiasdanne) [inline week3/tideman/tideman.c]:
```suggestion
    }

    for (int i = 0; i < candidate_count; i++)
```

**Review by matthiasdanne — APPROVED** [review]
All in all this is good. 👍

I have tested the programs and they're working as expected.

File: week3/plurality/plurality.c [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        return 1;
+    }
+
+    candidate_count = argc - 1;
+    if (candidate_count > MAX)
+    {
+        printf("Maximum number of candidates is %i\n", MAX);
+        return 2;
+    }
+    for (int i = 0; i < candidate_count; i++)
```
Comment (stefanluth) [inline week3/plurality/plurality.c]:
```suggestion
    }

    for (int i = 0; i < candidate_count; i++)
```

File: week3/runoff/runoff.c [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+
+int preferences[MAX_VOTERS][MAX_CANDIDATES];
+
+typedef struct
+{
+    string name;
+    int votes;
+    bool eliminated;
+}
+candidate;
```
Comment (stefanluth) [inline week3/runoff/runoff.c]:
```suggestion
} candidate;
```

File: week3/runoff/runoff.c [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+
+int candidate_count;
+int voter_count;
+
+void eliminate(int min);
+int find_min(void);
+bool is_tie(int min);
+bool print_winner(void);
+void tabulate(void);
+bool vote(int voter, int rank, string name);
```
Comment (stefanluth) [inline week3/runoff/runoff.c]: Sort by types, then alphabetically.

File: week3/runoff/runoff.c [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+        return 1;
+    }
+
+    candidate_count = argc - 1;
+    if (candidate_count > MAX_CANDIDATES)
+    {
+        printf("Maximum number of candidates is %i\n", MAX_CANDIDATES);
+        return 2;
+    }
+    for (int i = 0; i < candidate_count; i++)
```
Comment (stefanluth) [inline week3/runoff/runoff.c]:
```suggestion
    }

    for (int i = 0; i < candidate_count; i++)
```

File: week3/runoff/runoff.c [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+    voter_count = get_int("Number of voters: ");
+    if (voter_count > MAX_VOTERS)
+    {
+        printf("Maximum number of voters is %i\n", MAX_VOTERS);
+        return 3;
+    }
+
+    for (int i = 0; i < voter_count; i++)
+    {
+
```
Comment (stefanluth) [inline week3/runoff/runoff.c]:
```suggestion
```

File: week3/tideman/tideman.c [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+int candidate_count;
+int pair_count;
+
+void add_pairs(void);
+bool hasCycle(int winner, int loser);
+void lock_pairs(void);
+void print_winner(void);
+void record_preferences(int ranks[]);
+void sort_pairs(void);
+bool vote(int rank, string name, int ranks[]);
```
Comment (stefanluth) [inline week3/tideman/tideman.c]: s.o.

File: week3/tideman/tideman.c [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+        record_preferences(ranks);
+
+        printf("\n");
+    }
+
+    add_pairs();
+    sort_pairs();
+    lock_pairs();
+    print_winner();
+    return 0;
```
Comment (stefanluth) [inline week3/tideman/tideman.c]:
```suggestion
    print_winner();

    return 0;
```

File: week3/plurality/plurality.c [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    print_winner();
+}
+
+bool vote(string name)
+{
+    for (int i = 0; i < candidate_count; i++)
+    {
+        bool NameEqualName = strcmp(candidates[i].name, name) == 0;
```
Comment (stefanluth) [inline week3/plurality/plurality.c]:
```suggestion
        bool NameEqualName = strcmp(candidates[i].name, name) == 0;
```
[Snake case](https://en.wikipedia.org/wiki/Snake_case), maybe `canditate_exists`

Comment (stefanluth) [reply week3/plurality/plurality.c]: In `runoff.c` etc as well.

File: week3/runoff/runoff.c [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+        }
+    }
+    return 0;
+}
+
+bool vote(int voter, int rank, string name)
+{
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (strcmp(candidates[i].name, name) == 0)
```
Comment (stefanluth) [inline week3/runoff/runoff.c]:
```suggestion
        bool variable = strcmp(candidates[i].name, name) == 0;
        if (variable)
```

File: week3/runoff/runoff.c [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+    
+    return false;
+}
+
+int find_min(void)
+{
+    int min = voter_count;
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (candidates[i].votes < min && candidates[i].eliminated == false)
```
Comment (stefanluth) [inline week3/runoff/runoff.c]:
```suggestion
        bool variable = candidates[i].votes < min && candidates[i].eliminated == false;
        if (variable) 
```

File: week3/runoff/runoff.c [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+    return min;
+}
+
+bool is_tie(int min)
+{
+    int eliminate = 0;
+    int counter = 0;
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if (!candidates[i].eliminated)
```
Comment (stefanluth) [inline week3/runoff/runoff.c:178]:
```suggestion
        if (!candidates[i].eliminated)
```

Sometimes you use `!eliminated` and other times `eliminated == false`. Choose one

File: week3/runoff/runoff.c [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+        {
+            eliminate++;
+        }
+        if (candidates[i].votes == min)
+        {
+            counter++;
+        }
+    }
+    
+    return false;
```
Comment (stefanluth) [inline week3/runoff/runoff.c]: This always returns false, is this intended?

File: week3/tideman/tideman.c [hunk 27]
```diff
@@ -0,0 +1,229 @@
...
+    return;
+}
+
+void add_pairs(void)
+{
+    for (int i = 0; i < candidate_count; i++)
+    {
+        for (int j = i + 1; j < candidate_count; j++)
+        {
+            if (preferences[i][j] != preferences[j][i])
```
Comment (stefanluth) [inline week3/tideman/tideman.c]:
```suggestion
            bool variable = preferences[i][j] != preferences[j][i];
            if (variable)
```

File: week3/tideman/tideman.c [hunk 28]
```diff
@@ -0,0 +1,229 @@
...
+
+    return;
+}
+
+void sort_pairs(void)
+{
+    for (int i = 0; i < pair_count; i++)
+    {
+        int max_index = i;
+        int current_strenght = preferences[pairs[i].winner][pairs[i].loser] - preferences[pairs[i].loser][pairs[i].winner];
```
Comment (stefanluth) [inline week3/tideman/tideman.c]:
```suggestion
        int current_strength = preferences[pairs[i].winner][pairs[i].loser] - preferences[pairs[i].loser][pairs[i].winner];
```

File: week3/tideman/tideman.c [hunk 29]
```diff
@@ -0,0 +1,229 @@
...
+
+        for (int j = i + 1; j < pair_count; j++)
+        {
+            int temp_strenght = preferences[pairs[j].winner][pairs[j].loser] - preferences[pairs[j].loser][pairs[j].winner];
```
Comment (stefanluth) [inline week3/tideman/tideman.c]:
```suggestion
            int temp_strength = preferences[pairs[j].winner][pairs[j].loser] - preferences[pairs[j].loser][pairs[j].winner];
```

File: week3/tideman/tideman.c [hunk 30]
```diff
@@ -0,0 +1,229 @@
...
+                max_index = j;
+                current_strenght = preferences[pairs[j].winner][pairs[j].loser] - preferences[pairs[j].loser][pairs[j].winner];
+            }
+        }
+
+        pair temp = pairs[max_index];
+        pairs[max_index] = pairs[i];
+        pairs[i] = temp;
+    }
+}
```
Comment (stefanluth) [inline week3/tideman/tideman.c:167]:
```suggestion
    }
    
    return;
}
```

**Review by stefanluth — CHANGES_REQUESTED** [review]
Only reviewed code style. Try to keep using one naming convention like snake_case and add use variables for more complicated if-checks.

File: week3/runoff/runoff.c [hunk 31]
```diff
@@ -0,0 +1,202 @@
...
+        return false;
+}
+
+void tabulate(void)
+{
+    for (int i = 0; i < voter_count; i++)
+    {
+        for (int j = 0; j < candidate_count; j++)
+        {
+            if (candidates[preferences[i][j]].eliminated == false)
```
Comment (stefanluth) [inline week3/runoff/runoff.c]:
```suggestion
            if (!candidates[preferences[i][j]].eliminated)
```

File: week3/runoff/runoff.c [hunk 32]
```diff
@@ -0,0 +1,202 @@
...
+    
+    return false;
+}
+
+int find_min(void)
+{
+    int min = voter_count;
+    for (int i = 0; i < candidate_count; i++)
+    {
+        bool current_candidate_with_lowest_votes = candidates[i].votes < min && candidates[i].eliminated == false;
```
Comment (stefanluth) [inline week3/runoff/runoff.c]:
```suggestion
        bool current_candidate_with_lowest_votes = candidates[i].votes < min && !candidates[i].eliminated;
```

File: week3/tideman/tideman.c [hunk 33]
```diff
@@ -0,0 +1,233 @@
...
+    int loser;
+} pair;
+
+string candidates[MAX];
+pair pairs[MAX * (MAX - 1) / 2];
+
+int candidate_count;
+int pair_count;
+
+bool hasCycle(int winner, int loser);
```
Comment (stefanluth) [inline week3/tideman/tideman.c]: snake case

File: week3/tideman/tideman.c (hunk 29 above)
Comment (stefanluth) [reply week3/tideman/tideman.c]: Typo

File: week3/tideman/tideman.c [hunk 34]
```diff
@@ -0,0 +1,233 @@
...
+            }
+        }
+
+        pair temp = pairs[max_index];
+        pairs[max_index] = pairs[i];
+        pairs[i] = temp;
+    }
+}
+
+bool hasCycle(int winner, int loser)
```
Comment (stefanluth) [inline week3/tideman/tideman.c]: snake case

**Review by stefanluth — CHANGES_REQUESTED** [review]

**Review by stefanluth — APPROVED** [review]

**Pull request merged**